code loading: more helpful error messages

StefanKarpinski opened the pull request.

StefanKarpinski committed:
code loading: more helpful error messages (mostly fix #26472)

quinnj commented:
Yay!

KristofferC commented on base/loading.jl:
```diff
@@ -865,8 +865,26 @@ For more details regarding code loading, see the manual.
+        where = PkgId(into)
+        if where.uuid === nothing
+            throw(ArgumentError("""
+                Package $mod not found in current path:
```
Suggesting just `Pkg.add` would perhaps make the message more light weight.

staticfloat replied:
Alternatively, you could say `not found in current LOAD_PATH`, to give the user a hint that such a thing exists and close the loop at bit for otherwise clueless users.

KristofferC commented on base/loading.jl:
```diff
@@ -947,7 +965,11 @@ function _require(pkg::PkgId)
         path = locate_package(pkg)
         if path === nothing
-            throw(ArgumentError("Module $name not found in current path.\nRun `Pkg.add(\"$name\")` to install the $name package."))
+            throw(ArgumentError("""
```
Again, perhaps just recommend `Pkg.instantiate()` instead of the repl.

KristofferC replied:
For the future we should also check if we failed to find an explicit path in the manifest  and print that one but could be added later.

KristofferC reviewed (commented):
I think we should use the API interface in the error messages since those can be copy pasted and it makes the error messages shorter.

miguelraz commented on base/loading.jl:
```diff
@@ -947,7 +965,11 @@ function _require(pkg::PkgId)
+            throw(ArgumentError("""
+                Package $pkg is required but does not seem to be installed:
+                 - Activate the `pkg>` prompt by pressing the `]` key.
+                 - Run `pkg> instantiate` to install all recorded dependencies.
```
Perhaps useful to note the file/path where these dependencies are recorded?

StefanKarpinski commented:
> I think we should use the API interface in the error messages since those can be copy pasted and it makes the error messages shorter.

But this is also a chance to teach people how to use the package REPL mode. Of course, perhaps they'll learn that anyway, but it seems better to instruct them to use it one way or the other.

KristofferC commented:
I think we can work on promoting the REPL  mode through other channels. Tutorial style error messages is in my opinion annoying in the long run (after the first time you see them).

StefanKarpinski commented:
So the instructions should be `import Pkg; Pkg.API.add("Foo")`? That seems embarrassingly verbose.

KristofferC commented:
No need for `API`,  no, it is literally the same command as before. Regarding having to import, we could re-export it from Base but that kinda goes against the stdlib mentality. We could perhaps have it load by default in the REPL like the InteractiveUtils stuff.

StefanKarpinski commented:
Currently, `Pkg.instantiate()` does call the new `Pkg.API` but it also prints a deprecation warning:
```jl
julia> Pkg.instantiate()
WARNING: Base.Pkg is deprecated, run `using Pkg` instead
 in module Main
  Updating registry at `~/.julia/registries/Uncurated`
  Updating git-repo `https://github.com/JuliaRegistries/Uncurated.git`
```
Exporting `Pkg` from `InteractiveUtils` is a better approach. We'd have to delete the current deprecated binding for `Base.Pkg` to `stdlib/OldPkg` but that's probably for the better anyway.

KristofferC commented:
>  but it also prints a deprecation warning

Yes, because Pkg is an stdlib. 

> We'd have to delete the current deprecated binding for Base.Pkg to stdlib/OldPkg

IIRC `Base.Pkg` binds to `Pkg`, not `OldPkg`. Otherwise `Pkg.instantiate` wouldn't work.

ararslan commented on base/loading.jl:
```diff
@@ -865,8 +865,26 @@ For more details regarding code loading, see the manual.
+                 - Activate the `pkg>` prompt by pressing the `]` key.
+                 - If you have $(where.name) checked out for development and have
+                   added $mod as a dependency but haven't updated your primary
+                   environment's manifest file, try doing `pkg> up --fixed` to
```
Seems there isn't a `Pkg.update(fixed=true)` or similar, but that would make a good addition that would simplify this as well.

KristofferC replied:
Yeah the API for setting upgrade on levels is currently not so user friendly

Perhaps there should just be a`resolve` that does this.

Also, arguably the message here should have the `--manifest` flag.

StefanKarpinski committed:
InteractiveUtils: export Pkg making it available at the REPL

Allows instructions like `Pkg.add("Foo")` in error messages.

People won't need to restart the REPL when they try to do what
an error message tells them to do without `import Pkg` first.

StefanKarpinski committed:
Pkg: add `Pkg.resolve()` API function

ararslan commented on base/sysimg.jl:
```diff
@@ -583,7 +582,6 @@ end
     @deprecate_binding REPLCompletions root_module(Base, :REPL).REPLCompletions true ", use `REPL.REPLCompletions` instead"
     @deprecate_binding Terminals       root_module(Base, :REPL).Terminals       true ", use `REPL.Terminals` instead"
 
-    @deprecate_binding Pkg root_module(Base, :Pkg) true ", run `using Pkg` instead"
```
Will doing this be a hard break for CI scripts that run things like `julia -e 'Pkg.thing()'`?

KristofferC replied:
Yes, I don't think we should do this until CI scripts are updated to import Pkg.

StefanKarpinski replied:
I don't think we can have a deprecated binding exported from Base and have `using InteractiveUtils` import `Pkg` at the same time. I guess I'll try to figure out a way to make that happen though.

StefanKarpinski replied:
Well, I'll be damned, it seems to work. I guess because both bindings are to the same thing.

ararslan commented:
It could be worthwhile to have a test that Pkg is available when InteractiveUtils is loaded.

StefanKarpinski commented:
I want to get this merged now while the CI is green and the PR is fresh, we can make wording tweaks later.
